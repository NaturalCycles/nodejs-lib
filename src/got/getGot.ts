import { URL } from 'url'
import { _since } from '@naturalcycles/js-lib'
import got, { AfterResponseHook, BeforeErrorHook, BeforeRequestHook, Got, HTTPError } from 'got'
import { inspectAny } from '..'
import { dimGrey, grey, red, yellow } from '../colors'
import { GetGotOptions, GotRequestContext } from './got.model'

/**
 * Returns instance of Got with "reasonable defaults":
 *
 * 1. Error handler hook that prints helpful errors.
 * 2. Hooks that log start/end of request (optional, false by default).
 * 3. Reasonable defaults(tm), e.g non-infinite Timeout
 */
export function getGot(opt: GetGotOptions = {}): Got {
  opt.logger ||= console

  return got.extend({
    // Most-important is to set to anything non-empty (so, requests don't "hang" by default).
    // Should be long enough to handle for slow responses from scaled cloud APIs in times of spikes
    // Ideally should be LESS than default Request timeout in backend-lib (so, it has a chance to error
    // before server times out with 503).
    timeout: 90_000,
    ...opt,
    hooks: {
      ...opt.hooks,
      beforeError: [
        ...(opt.hooks?.beforeError || []),
        // User hooks go BEFORE
        gotErrorHook(opt),
      ],
      beforeRequest: [
        gotBeforeRequestHook(opt),
        // User hooks go AFTER
        ...(opt.hooks?.beforeRequest || []),
      ],
      afterResponse: [
        ...(opt.hooks?.afterResponse || []),
        // User hooks go BEFORE
        gotAfterResponseHook(opt),
      ],
    },
  })
}

/**
 * Without this hook (default behaviour):
 *
 * HTTPError: Response code 422 (Unprocessable Entity)
 *  at EventEmitter.<anonymous> (.../node_modules/got/dist/source/as-promise.js:118:31)
 *  at processTicksAndRejections (internal/process/task_queues.js:97:5) {
 *  name: 'HTTPError'
 *
 *
 * With this hook:
 *
 * HTTPError 422 GET http://a.com/err?q=1 in 8 ms
 * {
 *   message: 'Reference already exists',
 *   documentation_url: 'https://developer.github.com/v3/git/refs/#create-a-reference'
 * }
 *
 * Features:
 * 1. Includes original method and URL (including e.g searchParams) in the error message.
 * 2. Includes response.body in the error message (limited length).
 * 3. Auto-detects and parses JSON response body (limited length).
 * 4. Includes time spent (gotBeforeRequestHook must also be enabled).
 * UPD: excluded now to allow automatic Sentry error grouping
 */
function gotErrorHook(opt: GetGotOptions = {}): BeforeErrorHook {
  const { maxResponseLength = 10_000 } = opt

  return err => {
    if (err instanceof HTTPError) {
      const { statusCode } = err.response
      const { method, url, prefixUrl } = err.options
      const shortUrl = getShortUrl(opt, url, prefixUrl)
      // const { started } = context as GotRequestContext

      const body = inspectAny(err.response.body, {
        maxLen: maxResponseLength,
        colors: false,
      })

      // timings are not part of err.message to allow automatic error grouping in Sentry
      err.message = [[statusCode, method, shortUrl].filter(Boolean).join(' '), body]
        .filter(Boolean)
        .join('\n')
    }

    return err
  }
}

function gotBeforeRequestHook(opt: GetGotOptions): BeforeRequestHook {
  return options => {
    options.context = {
      ...options.context,
      started: Date.now(),
    } as GotRequestContext

    if (opt.logStart) {
      const shortUrl = getShortUrl(opt, options.url, options.prefixUrl)
      opt.logger!.log([dimGrey(' >>'), dimGrey(options.method), grey(shortUrl)].join(' '))
    }
  }
}

function gotAfterResponseHook(opt: GetGotOptions = {}): AfterResponseHook {
  return resp => {
    const success = resp.statusCode >= 200 && resp.statusCode < 400

    if (opt.logFinished) {
      const { started } = resp.request.options.context as GotRequestContext
      const { url, prefixUrl, method } = resp.request.options
      const shortUrl = getShortUrl(opt, url, prefixUrl)

      opt.logger!.log(
        [
          dimGrey(' <<'),
          coloredHttpCode(resp.statusCode),
          dimGrey(method),
          grey(shortUrl),
          started && dimGrey('in ' + _since(started)),
        ]
          .filter(Boolean)
          .join(' '),
      )
      // console.log(`afterResp! ${resp.request.options.method} ${resp.url}`, { context: resp.request.options.context })
    }

    // Error responses are not logged, cause they're included in Error message already
    if (opt.logResponse && success) {
      opt.logger!.log(inspectAny(resp.body, { maxLen: opt.maxResponseLength }))
    }

    return resp
  }
}

function coloredHttpCode(statusCode: number): string {
  if (statusCode < 400) return dimGrey(statusCode) // default
  if (statusCode < 500) return yellow(statusCode)
  return red(statusCode)
}

function getShortUrl(opt: GetGotOptions, url: URL, prefixUrl?: string): string {
  let shortUrl = url.toString()

  if (opt.logWithSearchParams === false) {
    shortUrl = shortUrl.split('?')[0]!
  }

  if (opt.logWithPrefixUrl === false && prefixUrl && shortUrl.startsWith(prefixUrl)) {
    shortUrl = shortUrl.slice(prefixUrl.length)
  }

  return shortUrl
}
