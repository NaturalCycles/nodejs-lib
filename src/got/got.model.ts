export interface GetGotOptions extends GotBeforeRequestHookOptions, GotAfterResponseHookOptions {}

export interface GotErrorHookOptions extends GotMaxResponseLength {}

export interface GotBeforeRequestHookOptions {
  /**
   * @default false
   */
  logStart?: boolean
}

export interface GotAfterResponseHookOptions extends GotMaxResponseLength {
  /**
   * Log when request is finished.
   * @default false
   */
  logFinished?: boolean

  /**
   * Log actual response object.
   *
   * @default false
   */
  logResponse?: boolean
}

export interface GotMaxResponseLength {
  /**
   * Max length of response object before it's truncated.
   *
   * @default 1000
   */
  maxResponseLength?: number
}

export interface GotRequestContext {
  /**
   * Millisecond-timestamp of when the request was started. To be able to count "time spent".
   */
  started: number
}
