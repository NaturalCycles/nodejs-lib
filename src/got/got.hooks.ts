import { _isHttpErrorResponse, _jsonParseIfPossible, _since } from '@naturalcycles/js-lib'
import got, { AfterResponseHook, BeforeErrorHook, BeforeRequestHook, Got, HTTPError } from 'got'
import { inspectAny } from '..'
import { dimGrey, grey, red, yellow } from '../colors'
import {
  GetGotOptions,
  GotAfterResponseHookOptions,
  GotBeforeRequestHookOptions,
  GotErrorHookOptions,
  GotRequestContext,
} from './got.model'

/**
 * Returns instance of Got with "reasonable defaults":
 *
 * 1. Error handler hook that prints helpful errors.
 * 2. Hooks that log start/end of request (optional, false by default).
 */
export function getGot(opt: GetGotOptions = {}): Got {
  return got.extend({
    hooks: {
      beforeError: [gotErrorHook(opt)],
      beforeRequest: [gotBeforeRequestHook(opt)],
      afterResponse: [gotAfterResponseHook(opt)],
    },
  })
}

/**
 * Without this hook (default behaviour):
 *
 * HTTPError: Response code 422 (Unprocessable Entity)
 *  at EventEmitter.<anonymous> (.../node_modules/got/dist/source/as-promise.js:118:31)
 *  at processTicksAndRejections (internal/process/task_queues.js:97:5) {
 *  name: 'HTTPError'
 *
 *
 * With this hook:
 *
 * HTTPError 422 GET http://a.com/err?q=1 in 8 ms
 * {
 *   message: 'Reference already exists',
 *   documentation_url: 'https://developer.github.com/v3/git/refs/#create-a-reference'
 * }
 *
 * Features:
 * 1. Includes original method and URL (including e.g searchParams) in the error message.
 * 2. Includes response.body in the error message (limited length).
 * 3. Auto-detects and parses JSON response body (limited length).
 * 4. Includes time spent (gotBeforeRequestHook must also be enabled).
 */
export function gotErrorHook(opt: GotErrorHookOptions = {}): BeforeErrorHook {
  const { maxResponseLength = 10000 } = opt

  return err => {
    if (err instanceof HTTPError) {
      const { statusCode } = err.response
      const { method, url, context } = err.options
      const { started } = context as GotRequestContext

      // Auto-detect and prettify JSON response (if any)
      let body = _jsonParseIfPossible(err.response.body)

      // Detect HttpErrorResponse
      if (_isHttpErrorResponse(body)) {
        body = body.error
      }

      body = inspectAny(body, {
        maxLen: maxResponseLength,
        colors: false,
      })

      err.message = [
        [statusCode, method, url, started && `in ${_since(started)}`].filter(Boolean).join(' '),
        body,
      ]
        .filter(Boolean)
        .join('\n')
    }

    return err
  }
}

export function gotBeforeRequestHook(opt: GotBeforeRequestHookOptions = {}): BeforeRequestHook {
  return options => {
    options.context = {
      ...options.context,
      started: Date.now(),
    } as GotRequestContext

    if (opt.logStart) {
      console.log([dimGrey(' >>'), dimGrey(options.method), grey(options.url)].join(' '))
    }
  }
}

export function gotAfterResponseHook(opt: GotAfterResponseHookOptions = {}): AfterResponseHook {
  return resp => {
    const success = resp.statusCode >= 200 && resp.statusCode < 400

    if (opt.logFinished) {
      const { started } = resp.request.options.context as GotRequestContext

      console.log(
        [
          dimGrey(' <<'),
          coloredHttpCode(resp.statusCode),
          dimGrey(resp.request.options.method),
          grey(resp.request.options.url),
          started && dimGrey('in ' + _since(started)),
        ]
          .filter(Boolean)
          .join(' '),
      )
      // console.log(`afterResp! ${resp.request.options.method} ${resp.url}`, { context: resp.request.options.context })
    }

    // Error responses are not logged, cause they're included in Error message already
    if (opt.logResponse && success) {
      console.log(inspectAny(_jsonParseIfPossible(resp.body), { maxLen: opt.maxResponseLength }))
    }

    return resp
  }
}

function coloredHttpCode(statusCode: number): string {
  if (statusCode < 400) return dimGrey(statusCode) // default
  if (statusCode < 500) return yellow(statusCode)
  return red(statusCode)
}
