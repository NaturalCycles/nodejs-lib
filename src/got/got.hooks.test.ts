import { mockTime } from '@naturalcycles/dev-lib'
import * as nock from 'nock'
import { getGot } from './got.hooks'

beforeAll(() => {
  mockTime()
})

afterAll(() => {
  nock.cleanAll()
  nock.restore() // prevents jest memory leak
})

const okResp = { ok: 1 }
const errResp = {
  message: 'Reference already exists',
  documentation_url: 'https://developer.github.com/v3/git/refs/#create-a-reference',
}

nock(/.*/)
  .persist()
  .get(/.*/)
  .reply((uri, _body: any) => {
    // console.log({uri, _body})

    if (uri.includes('err')) {
      return [422, errResp]
    }

    return [200, okResp]
  })

const _got = getGot({
  logStart: true,
  logFinished: true,
  logResponse: true,
})

test('gotErrorHook', async () => {
  expect(await _got.get('http://a.com').json()).toEqual(okResp)

  await expect(
    _got.get('http://a.com/err', {
      searchParams: { q: 1 },
    }),
  ).rejects.toThrowErrorMatchingSnapshot()
})

test.skip('actual error', async () => {
  await _got.get('http://a.com/err', {
    searchParams: { q: 1 },
  })
})
