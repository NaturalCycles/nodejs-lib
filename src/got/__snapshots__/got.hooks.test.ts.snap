// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gotErrorHook 1`] = `
"422 GET http://a.com/err?q=1 in 0 ms
{
  message: 'Reference already exists',
  documentation_url: 'https://developer.github.com/v3/git/refs/#create-a-reference'
}"
`;
